### PR TITLE
[DeadCode] Fix variable usage detection

### PR DESCRIPTION
**Pull request by dobryy**
Closes https://github.com/rectorphp/rector/issues/3726

### CODE REVIEW

File: rules/dead-code/src/NodeFinder/NextVariableUsageNodeFinder.php [hunk 1]
```diff
@@ -80,4 +81,25 @@ public function find(Assign $assign): ?Node
 
         return $nextUsageOfVariable;
     }
+
+    private function findInParentExpression(Assign $assign, Variable $variable): bool
```
Comment (TomasVotruba) [inline rules/dead-code/src/NodeFinder/NextVariableUsageNodeFinder.php:85]: `find*` usually return object/array, e.g. `findProduct` returns `Product`

Here should be `is*/has*` prefix... 

I hear you, this should be already in PHPStan rules :thinking: 

<br>

A better position for this method might be `BetterNodeFinder`

File: rules/dead-code/tests/Rector/Assign/RemoveUnusedAssignVariableRector/Fixture/skip_assign_expression_class_use_variable.php.inc [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\DeadCode\Tests\Rector\Assign\RemoveUnusedAssignVariableRector\Fixture;
+
+final class SkipAssignExpressionClassUseVariable
+{
+    public function run(string $foo = null, string $bar = null)
```
Comment (TomasVotruba) [inline rules/dead-code/tests/Rector/Assign/RemoveUnusedAssignVariableRector/Fixture/skip_assign_expression_class_use_variable.php.inc:9]:
```suggestion
    public function run()
```

Test fixture case should be as minimal as possible to make it easier in 2 years to debug them.

From the code I assume this is never used to test as tested variable name is `datetime`

File: rules/dead-code/src/NodeFinder/NextVariableUsageNodeFinder.php (hunk 1 above)
Comment (dobryy) [reply rules/dead-code/src/NodeFinder/NextVariableUsageNodeFinder.php:85]: How often do you hear voices? :D

File: rules/dead-code/tests/Rector/Assign/RemoveUnusedAssignVariableRector/Fixture/skip_assign_expression_class_use_variable.php.inc (hunk 2 above)
Comment (dobryy) [reply rules/dead-code/tests/Rector/Assign/RemoveUnusedAssignVariableRector/Fixture/skip_assign_expression_class_use_variable.php.inc:9]: You're right, missed during some testing.

**Pull request merged**